app/Models/Concerns/HasMemberships.php

```diff
@@ -0,0 +1,148 @@
+    {
+        $matchingMembership = resolve(Membership::class)->findPrimaryByVATSIMLocality($division, $region);
+
+        if (! $matchingMembership) {
+            return false;
+        }
+
+        // Check if the user already has this membership, and the division/region combination is the same
+        if ($this->hasMembership($matchingMembership) && ($this->primaryMembership()->pivot->division == $division && $this->primaryMembership()->pivot->region == $region)) {
+            return false;
+        }
+
+        // Check we can have secondary memberships
+        if (! $matchingMembership->can_have_secondaries) {
+            $this->removeSecondaryMemberships();
+        }
+
+        // End Previous Primary Membership
+        if ($currentMembership = $this->primaryMembership()) {
+            $this->removeMembership($currentMembership);
+        }
+
+        $this->memberships()->attach($matchingMembership, [
+            'division' => $division,
+            'region' => $region,
+            'started_at' => Carbon::now(),
+        ]);
+
+        return true;
+    }
+
+    public function addSecondaryMembership(Membership $membership): bool
+    {
+        throw_if($membership->primary, new MembershipNotSecondaryException());
+
+        // Check we can have secondary memberships
+        if ($this->primaryMembership() && ! $this->primaryMembership()->can_have_secondaries) {
+            throw new PrimaryMembershipDoesntAllowSecondaryException();
+        }
+
+        // Check if the user already has this membership
+        if ($this->hasMembership($membership)) {
+            return false;
+        }
+
+        $this->memberships()->attach($membership);
+
+        return true;
+    }
+
+    public function removeSecondaryMemberships(): void
+    {
+        $this->secondaryMemberships->each(function ($membership) {
+            $this->removeMembership($membership);
+        });
+    }
+
+    /**
+     * @param Membership|int|string $membership
+     * @return bool
+     */
+    public function hasMembership($membership): bool
+    {
+        if ($membership instanceof Membership) {
+            return $this->memberships()->where('memberships.id', $membership->id)->exists();
+        } elseif (is_string($membership)) {
+            // Assume is an identifier
+            return $this->memberships()->where('memberships.identifier', $membership)->exists();
+        }
+        // Assume is an ID
+        return $this->memberships()->where('memberships.id', $membership)->exists();
+    }
+
+    public function removeMembership(Membership $state): int
+    {
+        return $this->memberships()->updateExistingPivot($state, [
+            'ended_at' => Carbon::now(),
+        ]);
+    }
+
+    public function getPrimaryMembershipAttribute(): ?Membership
+    {
+        return $this->primaryMembership();
+    }
+
+    public function getIsHomeMemberAttribute(): bool
```

Comment:
I'm not a fan of these methods, can you not just have a `getIsMemberType($memberType)` method?

Reply:
I did this because once we are consuming the API with other services, we don't want to hard code the DIV, INT etc. identifiers into each application. This allows auth to determine what qualifies for each membership and will remove faffery needed on the consuming end